go/teams/seitan.go

```diff
@@ -0,0 +1,152 @@
+package teams
+
+import (
+	//"fmt"
+	"bytes"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	libkb "github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"golang.org/x/crypto/nacl/secretbox"
+	"golang.org/x/crypto/scrypt"
+	"golang.org/x/net/context"
+
+	"github.com/keybase/saltpack/encoding/basex"
+)
+
+const SeitanRawIKeyLength = 10
+const SeitanEncodedIKeyLength = 16
+
+// Key-Base 34 encoding. lower case letters and digits except for 0 and 1.
+const KBase34EncodeStd = "abcdefghijklmnopqrstuvwxyz23456789"
+
+var Base34Encoding = basex.NewEncoding(KBase34EncodeStd, SeitanRawIKeyLength, "")
+
+// "Invite Key"
+type SeitanIKey string
+
+// "Packed Encrypted Invite Key"
+type SeitanPEIKey struct {
+	_struct           bool `codec:",toarray"`
+	Version           uint
+	TeamKeyGeneration keybase1.PerTeamKeyGeneration
+	RandomNonce       keybase1.BoxNonce
+	EIKey             []byte
+}
+
+func GenerateIKey() (ikey SeitanIKey, err error) {
+	rawKey, err := libkb.RandBytes(SeitanRawIKeyLength)
+	if err != nil {
+		return ikey, err
+	}
+
+	var encodedKey [SeitanEncodedIKeyLength]byte
+	Base34Encoding.Encode(encodedKey[:], rawKey)
+
+	var verify [10]byte
+	_, err = Base34Encoding.Decode(verify[:], encodedKey[:])
+	if err != nil {
+		return ikey, err
+	}
+
+	if !bytes.Equal(verify[:], rawKey) {
+		return ikey, errors.New("Internal error - ikey encoding failed")
+	}
+
+	ikey = SeitanIKey(encodedKey[:])
+	return ikey, nil
+}
+
+// "Stretched Invite Key"
+type SeitanSIKey [32]byte
+
+func (ikey SeitanIKey) GenerateSIKey() (sikey SeitanSIKey, err error) {
+	ret, err := scrypt.Key([]byte(ikey), nil, 4, 8, 1, 32)
+	if err != nil {
+		return sikey, err
+	}
+	if len(ret) != 32 {
+		return sikey, errors.New("internal error - scrypt did not return 32 bytes")
+	}
+	copy(sikey[:], ret)
+	return sikey, nil
+}
+
+func (sikey SeitanSIKey) GenerateTeamInviteID() (id SCTeamInviteID, err error) {
+	const SeitanInviteIDPayload = "{\"stage\": \"invite_id\"}"
+	const SeitanInviteIDSuffix = 0x0d
+
+	mac := hmac.New(sha256.New, sikey[:])
```

Comment:
i think in other places we're using SHA512.  let me change the iced to be the same.

Reply:
I think it is 512, I used 256 by mistake, thanks for catching this

Reply:
👍